src/sage/rings/finite_rings/integer_mod_ring.py

```diff
@@ -1547,6 +1551,418 @@ def random_element(self, bound=None):
+            sage: R.zero().roots()
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: root finding with multiplicities for this polynomial not implemented (try the multiplicities=False option)
+            sage: R.zero().roots(multiplicities=False)
+            [0, 1, 2, 3, 4, 5, 6]
+
+            sage: R.<x> = Zmod(8)[]
+            sage: R.zero().roots()
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: root finding with multiplicities for this polynomial not implemented (try the multiplicities=False option)
+            sage: R.zero().roots(multiplicities=False)
+            [0, 1, 2, 3, 4, 5, 6, 7]
+
+        This method doesn't support root-finding over rings that aren't Z/nZ:
+
+            sage: R.<x> = Zmod(120)[]
+            sage: f = x^2 - 1
+            sage: f.base_ring()._roots_univariate_polynomial(f, ring=RR, multiplicities=False)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError
+
+        Sage allows us to coerce polynomials from one modulus to another,
+        and that makes the following defined:
+
+            sage: R.<x> = Zmod(100)[]
+            sage: (x^2 - 1).roots(Zmod(99), multiplicities=False) == (x^2 - 1).change_ring(Zmod(99)).roots(multiplicities=False)
+            True
+        """
+
+        # This function only supports roots in an IntegerModRing
+        if ring is not self and ring is not None:
+            raise NotImplementedError
+
+        deg = f.degree()
+
+        if multiplicities:
+            if deg < 0 or not self.is_field():
+                raise NotImplementedError(
+                    "root finding with multiplicities for this polynomial not"
+                    " implemented (try the multiplicities=False option)"
+                )
+            # Roots of non-zero polynomial over finite fields by factorization
+            return f._roots_from_factorization(f.factor(), multiplicities)
+
+        # Zero polynomial is a base case
+        if deg < 0:
+            # All residue classes are roots of the zero polynomial
+            return [*map(self, range(self.cardinality()))]
+
+        # Finite fields are a base case
+        if self.is_field():
+            return f._roots_from_factorization(f.factor(), False)
+
+        # Otherwise, find roots modulo each prime power
+        fac = self.factored_order()
+        prime_power_roots = []
+        for p, e in fac:
+            Zpe = Zmod(p**e)
+            fpe = f.change_ring(Zpe)
+            fpe_prime = fpe.derivative()
+            fp = fpe.change_ring(Zmod(p))
+
+            mod_p_roots = fp.roots(multiplicities=False)
+
+            this_prime_power = []
+            prime_power_roots.append(this_prime_power)
+            for root in mod_p_roots:
+                this_prime_power.extend(
+                    self._lift_residue_field_root(p, e, fpe, fpe_prime, root)
+                )
```

Comment:
```suggestion
            this_prime_power = []
            for root in mod_p_roots:
                this_prime_power.extend(
                    self._lift_residue_field_root(p, e, fpe, fpe_prime, root)
                )
            prime_power_roots.append(this_prime_power)
```

I think this is more readable. I got tripped up reading this because the current code looks like you're appending an empty list and then doing nothing with `this_prime_power`, then I remembered that there's pointer/reference stuff going on.

Reply:
I'm in the habit of doing this kind of append immediately so that I don't forget it entirely. But I agree that it might be confusing, so it seems like a worthwhile change.